### DISCUSSION TITLE
Two Phase Waveform

### DISCUSSION THREAD

**Discussion by mosqu-ito** [SpenceKonde/DxCore#341] [Show and tell]
I've included fully _five_ decoupling capacitors for the sole purpose of galling @Warren52

![ManSized](https://user-images.githubusercontent.com/106849308/189003014-f6f0fdf0-45c2-44d4-86ae-78da0a0f4404.jpg)

<pre>
#include "MyAliases64DB28.h"
#include "MyGeneral64DB28.h"
#include "MyDac64DB28.h"
#include "MyOpAmps64DB28.h"
#include "SineData.h"           // 1024 words in flash

word Index = 0;

void setup() {
   DacSetup(VREF_REFSEL_VDD_gc);              // Vref,        ToPin (default true)    WHITE
   UnityGainFollower(FROM_DAC,          0);   // Input, Amp#, ToPin (default true)    YELLOW
   UnityGainInverter(FROM_PREVIOUS_AMP, 1);   // Input, Amp#, ToPin (default true)    BLUE
   }

void loop() { WriteDacRaw(pgm_read_word(&Sine[(Index++)&0x03FF])); }
</pre>
![TwoPhase](https://user-images.githubusercontent.com/106849308/189003177-732c0551-c34a-4e8e-971b-88525ea0257f.jpg)

**Comment by ObviousInRetrospect** [comment]
Huh the DB28 does come in dip. I always wondered why the heck anyone would use DB28s (other than the chip shortage) - in tssop they take up about the same board area as the 32's are somewhat gimped, and are more annoying to solder than the 0.8 pitch TQFP. Makes sense if it caters to the DIP crowd. Personally when I want a DIP-like mcu (rare, I hate solder less breadboards) I just mount a 32 on a breakout and solder on 2 0.1 headers.

**Reply by mosqu-ito** [reply]
I've got a couple of 48s but I'm still working up the courage to try to solder them to breakouts.

**Reply by ObviousInRetrospect** [reply]
if you don't have a lot of experience with SMT the TQFP-32's are much more forgiving. The 48 is a 0.5mm pitch, the 32 is a 0.8mm pitch. With the finer pitch you need to be accurate and bridging is harder to avoid. Lots of flux, fine leaded solder, and good solder wick help. Also you need some sort of magnification. If you don't have a microscope something like https://www.amazon.com/Illuminated-Microscope-Magnifiers-Magnifying-Magnification/dp/B08T9RV48H/ref=sr_1_24?tag=gr32-20&crid=1LR2H3IV26R3L&keywords=lighted+jewelers+loupe+60x&qid=1662614765&sprefix=lighted+jewelers+loupe+60%2Caps%2C181&sr=8-24 works acceptably (to inspect and find bridges). Tweezers are essential (for positioning and holding parts). 

But if you can get your hands on a few 32's (DD32s might still be easy to get) to try first, those are pretty easy and let you get a hang of it.

If you are totally new to SMT start with SOIC-14 (ATtiny3214). Those are 1.27mm pitch and you really have to try hard (or use ROHS solder) to create a bridge.

https://www.amazon.com/gp/product/B004X4L076/?tag=gr32-20&ref=ppx_yo_dt_b_search_asin_title is awesome (other suppliers are often cheaper)

https://www.amazon.com/gp/product/B008O9VLA2/?tag=gr32-20&ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1 is reasonable

https://www.amazon.com/iFixit-Precision-Tweezers-Set-Angled/dp/B079K874CQ/ref=sr_1_2?tag=gr32-20&crid=D2RNWO643BH7&keywords=fixit+tweezer+set&qid=1662614875&sprefix=fixit+tweezer+set%2Caps%2C175&sr=8-2 are a reasonable and cheap set of tweezers (the angled one gets the most use).

https://www.amazon.com/Chip-Quik-SMD291ST2CC6-Squeeze-SMD291/dp/B00HKJQ5P8/ref=sr_1_6?tag=gr32-20&crid=M8FTYVT3HI5I&keywords=chip+quik+flux&qid=1662615167&sprefix=chipquik+flux%2Caps%2C181&sr=8-6 is a reasonable flux for making it easier to avoid bridges. It also helps hold the part in place slightly.

You don't necessarily need all of the above but such things make it easier. I'm still on a roll of that solder I bought in 2016 with frequent use.

**Reply by mosqu-ito** [reply]
Thanks, OIR!  I am a big fan of Kester.  I'm not normally a brand loyalist, but I am when it comes to solder.  I've done SOIC 14, 16, 20, but as you say, it's a big step to the 48s.  And it doesn't help that I'm old, my house is littered with coke-bottle reading glasses and triplet handlenses just so I can tell the difference between naproxen and viagra.

**Reply by SpenceKonde** [reply]
Yeah. Truth be told, I don't even bother with western brands for solder >.> I use el cheapo stuff from aliexpress. I haven't done a deep study of wire solders (they're all the same except when they lie about their composition,. I've gotten "SAC305" that the seller then admitted was actually pure tin. It was worthless). I've put a ton more thought into solder past and spent proabably close to $200 on experimental solder pastes now, searching for a non-shitty lead free solder that melts at lower temperatures than SAC305 (my reflow oven doesn't get hot enough :-( ) For plain leaded solder, RELIFE is excellent, and the 181C silver-bearing KEK even better. For SAC305 unleaded I use KEK brand as well. The best intermediate temperature lead free ones I've found so far are PPD S600 (claimed 183 melting point and lead free), and 158C RELIFE/GLON ROHS solder. I've got a few more coming in the mail - PPD has a 158C one I need to try, and I found some randomass brand called "Laster" with 183 and 158C lead free solder pastes at a killer price. I'll probably get what I pay for there, but I'm not paying that much so w/e, and Wylie also makes a 158C one that I need to take for a spin. Most of the brands I've tried that were between 150C and 190C claimed melting point were really crappy SnBiAg alloys that yielded dull, lousy joints and had high failure rates. 

For flux, I use UGain UG78 for most tasks because it cleans up easier, but keep a tube of Kester RF741 on hand for when shit gets real. 

Oh, and you want either a knife tip or a tinned-face-only bevel for the soldering iron to drag solder wit the best possible results. The knife tip is incredible, because you can flip it around for precise work with the point, or use it on it's side for drag soldering. 

IMHO the hard part about the 48's isn't deaing with bridges, it's getting them aligned perfectly, because there are 5 axes along which it can be misaligned (2 translation, 1 rotation, and 2 more for offset of the center point of the rotation), and it can be fine on 3 sides and fucked on the fourth. so you have to be very careful about that. While larger QFNs get easier to solder (via hotplate or reflow) as the self-centering effect of the exposed pad becomes more prominent, QFPs get harder the larger they are, because the larger they are the smaller the angle of misalignment they can tolerate before one of the sides is off by half a pitch resulting in the pins briding the pads on either side, resulting in an entire side of pins shorted to each other,and you need hot air or a hotplate to fix it. .

**Reply by ObviousInRetrospect** [reply]
> I haven't done a deep study of wire solders (they're all the same except when they lie about their composition

They are certainly more similar than pastes. But the flux to solder ratio varies as does the type and quality of the flux. As well as the diameter. I have a few on my bench. They are most certainly different and for more demanding stuff it is very frustrating when I try to use (say) a thick no-clean mostly tin ROHS solder to solder 0.5mm pitch SMT parts. 

The solder I linked above is on the very high flux side and that flux (RA 44) is on the as active as you can get while still being able to sort of treat it as no-clean and not horrible to remove (50/50 mix of acetone and IPA in a jar in an ultrasonic bath takes it right off). It also spreads better than most. I do find that when soldering 0.5mm pitch fine solder gives me more control and even modestly thicker solder (0.020 vs 0.015) is far worse. Also 63/37 is much better than 60/40. I tend to do things with an iron that most people think can or should only be done with a stencil and paste because I hate dealing with paste cleanup and my volumes don't usually justify the stencil. So when pushing my limits I find using appropriate solder makes it easier. Although having figured out that you can get JLC to shrink the stencil so shipping isn't so outrageous I might do more that way. Need to figure out a good way to panelize stencil orders too at which point they would become cost effective.

Re alignment yeah it takes longer but I tend to solder under a microscope and with enough magnification you can tell when it is aligned. If your eyesight is poor a microscope may be worth the investment. (binocular with large working distance and high eye point if you wear glasses). 

Interesting point about larger QFNs being easier. I tend to do 16-24's and find them annoying because of the 5-axis issue. But I've taken to doing a quick and dirty job to apply some solder and flux and then hit it with hot air to center it properly.

And yeah cheapo hot-air rework stations ($40) work well enough. Much as I wish I still had a pencil.

@mosqu-ito your list includes SOICs but not any TQFPs. I'd suggest practicing on/mastering 0.8mm pitch TQFP 32's before stepping up to the 0.5mm ones. The 5-dimmensional alignment issue takes some practice. After the first few I expect 0.8mm pitch as pretty much guaranteed to go down correctly on the first try without rework and in 90-120 seconds for a 32 pin and 0.5s to be a bit fiddly and often require cleanup.

**Comment by Warren52** [comment]
Aw gee @mosqu-ito - you didn't have to go to all this effort!

Say, what do you have in mind for this Swiss army knife? Just out learning? 
Interesting chip!

**Reply by mosqu-ito** [reply]
I long ago stopped kidding myself that I was ever doing anything productive.  It's a lot more fun once you've decided that fun is all the justification you need.  I had never seen a microcontroller with onboard op amps before.  I was just keen to take them for a spin.

**Comment by mosqu-ito** [comment]
In my last post I was pontificating on the importance of "fun".  This, was not that.  I once spent all day passing several kidney stones; that was more fun than this.  If I ever express an intention to try this again, just shoot me.
![finally](https://user-images.githubusercontent.com/106849308/192152469-66e2a0c9-64fd-4465-a832-2c13221d7607.jpg)

**Reply by ObviousInRetrospect** [reply]
Looks legit. Except for the absence of decoupling caps. Here is my normal technique (when I don't just have osh make me a cheap BOB with the correct cap pads):

![image](https://user-images.githubusercontent.com/42229858/192158571-8395d02f-cffd-4bd3-addc-e62492c32525.png)

To do this you just scratch the solder mask off of the power trace, tin it, and solder a smt cap to it. I tend to just use the 0603s I use for everything else but in this case a bigger cap is easier as you are less likely to heat up the whole cap enough adding the wire that you desolder it from the trace.

Easiest with enameled magnet wire although 30awg works fine too. The silicone insulated stuff is less annoying than the pvc stuff. you can also solder leaded caps between the backs of the 0.1 headers (or the chip's pins if you enjoy torture).

And yes I think anyone who pins out chips with power and ground pretty much as far apart as possible is an idiot.

Doing fine pitch parts gets hugely easier with practice. I had a similar reaction to my first encounters with 0.5mm pitch parts. Then I discovered the proper types and use of flux.

I still think 0.8mm pitch TQFP32's are a cake-walk compared to anything 0.5mm. Almost never need either flux or wick on an 0.8. Usually start with flux and still need wick half the time on big 0.5s. If you are going to do them more than rarely its worth investing in a cheap hot air rework station. Perfectly serviceable ones can be had under $50. Makes it easy to correct your mistakes and sometimes fixing a bad job works by just melting the whole chip and tapping it into place. Worst case you melt, remove, wick, and re-apply.

**Reply by SpenceKonde** [reply]
They make magnet wire that's solderable now. The insulation burns off at soldering temperature, but is otherwise as thin and durable as normal enamel 

QFP48 is a completely different beast from QFP32, absolutely. You've got to admit another advantage of QFN. How many times have you bent the pin on a QFN? it's damned easy with 0.5mm pitch pins.

**Comment by mosqu-ito** [comment]
Three, just so I can have one more than you.  They just go wherever they look nice, right?

;)

Thanks for all the help and encouragement, OIR.
![looknice](https://user-images.githubusercontent.com/106849308/192163348-558a1612-6ccb-4411-a3f3-f3d40e82a361.jpg)

**Reply by SpenceKonde** [reply]
Those capacitors look suspiciously white to be 0.1uF...

**Reply by mosqu-ito** [reply]
You're very observant.  They're 103s, the 104s are similarly arranged on the socket pins under the board.  Very nearby I have one 105 dedicated to the MCU socket as a whole, and then one more 105 at every point where power enters or leaves the board.  Of course one can't get too carried away because if you get much above 10uF total, the plug-in surge can trip the overload on your USB port.

**Comment by mosqu-ito** [comment]
I won't make this a new thread under Q&A because I know others have had this kind of issue before; but when I went to program this, it complained that the Device ID was 1E9535 when it was expecting 1E9532.  Naturally I just went into AVR32DB48.py and changed it.  So far so good.  For what it's worth, the chip came from Digikey.

**Comment by SpenceKonde** [comment]
You did not do the correct thing. The correct thing to do was a two step process:
1. Download the latest copy of the AVR32DB datasheet, confirm that the device ID is supposed to be 1E9535 (it looks like it is is - if it was supposed to be 1E9532 on the other hand, step 2 would be to first verify the part number on the chip (if it's wrong, complain to digikey for shipping you the wrong thing and demand they ship you the product you ordered), and assuming that was correct, contact microchip support posthaste so that they can get the parts with wrong ID's recalled and rectified ASAP). But in this case, you determined (and I agree) that there is a defect in SerialUPDI, so the next step should be:
2. Create an issue here with a title like "SerialUPDI has incorrect device ID for the AVR32DB48 " (while your at it, you might look at the other AVR32DB's and see if they're also incorrect, but that would be done by me if you didn't). 

There sounds like there is not anything here to discuss, other than "How the fuck did you screw THAT up man? Don't you test your code?" (to which the answer is "no, of course not, I'm engineering, not quality assurance". I'll note that while Microchip swore up and down that we'd never have problems with stuff derived from pymcuprog working with new products because they'd keep everything up to date, I added those files myself, like 6 months after Microchip started shipping them and still hadn't updated it, and besides, the code as supplied was so dog slow as to be useless anyway, and after I made it perform acceptably, it was so far diverged from their tree that I couldn't merge in changes anyway.  So this is very much an issue, not a discussion. The only "discussion" that would be relevant would be insults directed at me. Which are of course always welcome - I used to play an MMO, at the top tiers of the game, so you have to really really work to make an insult that produces a response other than "That's a fair characterization, yes" or laughter

We are actually changing the name of the library so that pathological library updaters (the people who, without actually having done so themselves such that they'd know it wouldn't result in anything resembling functionality), instruct people on the forums to delete the pymcuprog directory we supply and install that basketcase from from microchip through pip) will be dissuaded from updating it and then bitching about all the problems they encounter

Discussions I have one chance to notice as they go past my email box, which sometimes receives so many emails in a day (most worthless) that it can go to over one page). The github issues section is my to-do list.

**Comment by mosqu-ito** [comment]
Finally, now that we're on a 48 pin DB, we have, not two, but three op amps.
Here's a simple straight-forward thing to do with the third one (OP2)...
<pre>
UnityGainFollower(HALF_VDD, 2);     // Input, Amp#, ToPin (default true)    GREEN
</pre>
Because the DBs bring us Op Amps, Microchip provides us with a voltage reference
they call "VDDDIV2" which I have here aliased as HALF_VDD.  There's actually not
a lot beyond the unity gain follower that you can do with an op amp without a third
rail (or external components).  Where there is no third rail you end up having to
concoct some approximation of one.  Hence the HALF_VDD reference.  It's great to
have but it's not exposed, and that leaves signals coming in/out from/to other
linear devices to deal with a DC offset - even your silly scope.  In my scope view
at the top you'll notice a little red arrow at the lower left.  That was me dragging off
the DC offset so the waveform wouldn't appear to be sitting way up on top of it.
Now, using the third op amp to stiffen and expose HALF_VDD, we have what you
might describe as a light-duty rail splitter, a reasonably usable linear ground from
which to reference op amp signals coming and going.
